components/remerge/src/untyped_map.rs

```diff
@@ -0,0 +1,300 @@
+                    log::trace!("UntypedMap tombstone collision for key {:?}", key);
+                    throw!(ErrorKind::UntypedMapTombstoneCollision)
+                }
+                CollisionHandling::DeleteEntry => {
+                    map.remove(&key);
+                }
+                CollisionHandling::KeepEntry => {
+                    tombs.remove(&key);
+                }
+            }
+        }
+        Ok(UntypedMap { map, tombs })
+    }
+
+    pub fn new<M, T>(map: M, tombstones: T, on_collision: OnCollision) -> Self
+    where
+        M: IntoIterator<Item = (String, JsonValue)>,
+        T: IntoIterator<Item = String>,
+    {
+        Self::new_impl(
+            map.into_iter().collect(),
+            tombstones.into_iter().collect(),
+            on_collision.into(),
+        )
+        .expect("bug: new_impl error when not passed CollisionHandling::Error")
+    }
+
+    pub fn try_new<M, T>(map: M, tombstones: T) -> Result<Self>
+    where
+        M: IntoIterator<Item = (String, JsonValue)>,
+        T: IntoIterator<Item = String>,
+    {
+        Self::new_impl(
+            map.into_iter().collect(),
+            tombstones.into_iter().collect(),
+            CollisionHandling::Error,
+        )
+    }
+
+    pub fn into_local_json(self) -> JsonValue {
+        serde_json::to_value(self).expect("UntypedMap can always be represented as json")
+    }
+
+    pub fn from_local_json(json: JsonValue) -> Result<Self> {
+        serde_json::from_value(json)
+            .map_err(Error::from)
+            .and_then(|Self { map, tombs }| {
+                // Ensure the entry is valid. TODO: eventually maintenance will
+                // need to handle this. Is this fine until then?
+                Self::try_new(map, tombs)
+            })
+    }
+
+    // Note: we don't use NativeRecord and such here since these are fields on
+    // records, and not actually records themselves. It's not really clear to me
+    // if/how we could use the RecordFormat markers or similar here either...
+    pub(crate) fn update_local_from_native(
+        old_local: JsonValue,
+        new_native: JsonValue,
+    ) -> Result<JsonValue> {
+        Ok(Self::from_local_json(old_local)?
+            .with_native(crate::util::into_obj(new_native)?)
+            .into_local_json())
+    }
+
+    /// Create a new representing an update of our data to the data in to
+    /// `new_native`, updating `tombs` in the process. Specifically:
+    /// 1. entries in `tombs` which refer to keys in `new_native` are removed
+    /// 2. entries in `self.map` which are missing from `new_native` are added
+    ///    to `tombs`.
+    #[must_use]
+    pub fn with_native(&self, new_native: JsonObject) -> Self {
+        let now_missing = self.map.keys().filter(|&k| !new_native.contains_key(k));
+
+        let tombs = now_missing
+            .chain(self.tombs.iter())
+            .filter(|t| !new_native.contains_key(t.as_str()))
```

Comment:
As an optimization, I think you could put the filter on `self.tombs`.

Reply:
There shouldn't be a large difference, and this way ensures correctness in the case that somehow things became inconsistent.